fix #5306 chore(nimbus): Parameterize publish_status with status_next

lmorchard opened the pull request:
Because:

* we want to distinguish between going through the publish cycle to
  update a live experiment and end an experiment

This commit:

* adds a new status_next field indicating the intended end target status
  for the current publishing flow in progress

* removes is_end_requested since status_next=Complete expresses the same

* migration and migration tests for is_end_requested and status_next

* updates GQL types and nimbus-ui frontend to use status_next

lmorchard commented:
Opening as draft because it seemed like there was one integration test not yet passing. But it passed locally, so it might be an intermittent failure? Will mark ready for review probably tomorrow if CI goes green

lmorchard commented:
Oh hey, maybe that failure *was* intermittent?

jaredlockhart commented on app/experimenter/experiments/api/v5/serializers.py:
```diff
@@ -458,12 +458,18 @@ def validate_secondary_outcomes(self, value):
-        if self.instance.status != NimbusExperiment.Status.LIVE:
+    def validate_status_next(self, value):
+        valid_status_next = NimbusExperiment.VALID_STATUS_NEXT_VALUES.get(
+            self.instance.status, ()
+        )
+        if value not in valid_status_next:
+            choices_str = ", ".join(str(choice) for choice in valid_status_next)
             raise serializers.ValidationError(
-                f"Nimbus Experiment has status '{self.instance.status}', but can only "
-                "be ended when set to 'Live'.",
+                f"Invalid choice for status_next: '{value}' - with status "
+                f"'{self.instance.status}', the only valid choices are "
+                f"'{choices_str}'"
             )
```
Very nice 👍

jaredlockhart commented on app/experimenter/experiments/constants/nimbus.py:
```diff
@@ -215,6 +215,12 @@ class PublishStatus(models.TextChoices):
     STATUS_ALLOWS_UPDATE = (Status.DRAFT,)
 
+    # Valid status_next values for given status values
+    VALID_STATUS_NEXT_VALUES = {
+        Status.DRAFT: (None, Status.LIVE),
+        Status.LIVE: (None, Status.LIVE, Status.COMPLETE),
+    }
```
👩‍🍳

jaredlockhart commented on app/experimenter/experiments/migrations/0174_switch_to_status_next.py:
```diff
@@ -0,0 +1,53 @@
+
+
+def convert_is_end_requested_to_status_next(apps, schema_editor):
+    NimbusExperiment = apps.get_model("experiments", "NimbusExperiment")
+    NimbusExperiment.objects.filter(
+        status=NimbusConstants.Status.LIVE, is_end_requested=True
+    ).update(status_next=NimbusConstants.Status.COMPLETE)
```
Good idea.  There's more cases though, aren't there?  

- Anything `DRAFT` that isn't `IDLE` should have `status_next=LIVE`
- Anything `LIVE` that isn't `IDLE` and has `is_end_requested=False` should have `status_next=LIVE`
- Anything `LIVE` that isn't `IDLE` and has `is_end_requested=True` should have `status_next=COMPLETE`

lmorchard replied:
Ah yeah, I think I tunnel-visioned on the is_end_requested case, will add these

jaredlockhart commented on app/experimenter/experiments/tests/factories/nimbus.py:
```diff
@@ -34,67 +34,76 @@
     }
     LIVE_IDLE_ENROLLING = {
         "status": NimbusExperiment.Status.LIVE,
+        # "status_next": NimbusExperiment.Status.LIVE,
```
Should be `None` here right?

jaredlockhart commented on app/experimenter/experiments/tests/factories/nimbus.py:
```diff
@@ -34,67 +34,76 @@
     }
     LIVE_WAITING_ENROLLING = {
         "status": NimbusExperiment.Status.LIVE,
+        # "status_next": NimbusExperiment.Status.LIVE,
```
This will be live when we do the next ticket but for now it's not used right?

jaredlockhart commented on app/experimenter/experiments/tests/test_migrations.py:
```diff
@@ -1,48 +1,99 @@
-from django.db.migrations.executor import MigrationExecutor
-from django.test.testcases import TransactionTestCase
+from django.utils.text import slugify
+from django_test_migrations.contrib.unittest_case import MigratorTestCase
```
The template we had didn't work for this?

lmorchard replied:
No, I fell down a rabbit hole of needing to access models from the pre-migration and post-migration states because of the change in fields. Things broke down when I tried to use the existing factories and the User model. This `django-test-migrations` package wrapped it all up pretty concisely

jaredlockhart replied:
Ah yep makes sense 👍

jaredlockhart reviewed (commented):
Very very nice @lmorchard 👍 

Couple little questions but generally everything looks exactly right, thank you!

jaredlockhart commented:
Oh also I was wondering whether we should track `status_next_old/new` in the changelogs.  It's technically getting tracked in the `experiment_data` field, but I'm wondering whether we'll need to filter on it later....  like to figure out launch/pause/update/end dates.  What do you think?

lmorchard commented:
Reworked the migration to account for more cases. Added status_next to the changelog model, but did *not* come up with a migration for that. Seemed like a bigger effort for data that probably wouldn't be immediately useful

jaredlockhart commented:
> Oh hey, maybe that failure _was_ intermittent?

Oh also yeah we do have an intermittent failure in the nimbus integration tests atm: https://github.com/mozilla/experimenter/issues/5415

If you see it again and it's not the same, pls also file it.

jaredlockhart commented on app/experimenter/experiments/changelog_utils/nimbus.py:
```diff
@@ -56,8 +58,10 @@ def generate_nimbus_changelog(experiment, changed_by, message):
+        old_status_next=old_status_next,
         old_publish_status=old_publish_status,
         new_status=experiment.status,
+        new_status_next=experiment.status_next,
```
👍 👍 👍 👍 👍

jaredlockhart reviewed (changes requested):
Just running through the flows locally and it looks like it gets stuck on launch timeout.

Steps to reproduce
1) Create experiment
2) Approve to launch to RS
3) Wait until timeout happens

Expected:
RS is rolled back
Experiment is in DRAFT/REVIEW(LIVE)

Actual:
RS is rolled back
Experiment is in 
Experiment is DRAFT/WAITING(LIVE)


So something got missed somewhere

jaredlockhart commented:
Oh interesting, I just filled it out as normal.  Let me try again.

lmorchard commented:
I see a local experiment with draft->waiting(live) that seems stuck which was generated as a dummy experiment, figured that's issue #5280 though

jaredlockhart commented:
Huh no, I just got the same, timed out in RS but no update in Experimenter.

I'll record a gif.

jaredlockhart commented:
Oh no, it just didn't work and then worked on the second pass.  That's even more annoying XD

lmorchard commented:
Couldn't reproduce the timeout issue, but in trying to do so I *did* find a different issue: Couldn't launch from Preview. Added another valid status_next transition and a test to cover that, pushed a new commit

lmorchard commented:
Ah ha, rebase reveals a new migration got in ahead of me. Gonna fix that and try CI again
